### PR TITLE
[yasm/xxx] Conan v2 migration

### PR DESCRIPTION
**Pull request by MartinDelille**
Specify library name and version:  **yasm/xxx**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/yasm/all/conanfile.py [hunk 1]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
+        tc.generate()
+        tc = PkgConfigDeps(self)
+        tc.generate()
+        tc = AutotoolsDeps(self)
+        tc.generate()
+        env = VirtualBuildEnv(self)
+        env.generate()
+        if not cross_building(self):
+            env = VirtualRunEnv(self)
+            env.generate(scope="build")
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:76]:
```suggestion
```

File: recipes/yasm/all/conanfile.py [hunk 2]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
+        return os.path.join(self.source_folder, "Mkfiles", "vc10")
+
+    def _generate_vs(self):
+        with chdir(self, self._msvc_subfolder):
+            tc = MSBuildToolchain(self)
+            tc.generate()
+            tc = MSBuildDeps(self)
+            tc.generate()
+            tc = VCVars(self)
+            tc.generate()
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:57]:
```suggestion
```

File: recipes/yasm/all/conanfile.py [hunk 3]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
-            self._build_vs()
+        if is_msvc(self):
+            msbuild = MSBuild(self)
+            msbuild.build_type = "Debug" if self.settings.build_type == "Debug" else "Release"
+            msbuild.platform = "Win32" if self.settings.arch == "x86" else msbuild.platform
+            msbuild.build(sln="project_2017.sln")
         else:
-            autotools = self._configure_autotools()
+            autotools = Autotools(self)
+            autotools.autoreconf()
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:92]:
```suggestion
```

File: recipes/yasm/all/conanfile.py [hunk 4]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
         del self.settings.compiler.cppstd
 
-    def build_requirements(self):
-        if self._settings_build.os == "Windows" and self.settings.compiler != "Visual Studio" and not tools.get_env("CONAN_BASH_PATH"):
-            self.build_requires("msys2/cci.latest")
+    def layout(self):
+        if is_msvc(self):
+            vs_layout(self)
+        else:
+            basic_layout(self, src_folder="src")
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:35]: discrepancy in source destination

File: recipes/yasm/all/conanfile.py [hunk 5]
```diff
@@ -1,9 +1,15 @@
...
+from conan.tools.gnu import Autotools, AutotoolsToolchain, AutotoolsDeps, PkgConfigDeps
+from conan.tools.layout import basic_layout, vs_layout
+from conan.tools.microsoft import is_msvc, MSBuildDeps, MSBuildToolchain, MSBuild, VCVars
+
 import os
-import shutil
 
-required_conan_version = ">=1.33.0"
 
+required_conan_version = ">=1.52.0"
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:12]: too high

File: recipes/yasm/all/conanfile.py [hunk 6]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
+            autotools.autoreconf()
+            autotools.configure()
             autotools.make()
 
     def package(self):
-        self.copy(pattern="BSD.txt", dst="licenses", src=self._source_subfolder)
-        self.copy(pattern="COPYING", dst="licenses", src=self._source_subfolder)
-        if self.settings.compiler == "Visual Studio":
+        copy(self, pattern="BSD.txt", dst="licenses", src=self.source_folder)
+        copy(self, pattern="COPYING", dst="licenses", src=self.source_folder)
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:98]:
```suggestion
        copy(self, pattern="BSD.txt", src=self.source_folder, dst=os.path.join(package_folder, "licenses"))
        copy(self, pattern="COPYING", src=self.source_folder, dst=os.path.join(package_folder, "licenses"))
```

File: recipes/yasm/all/conanfile.py [hunk 7]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
+        if is_msvc(self):
             arch = {
                 "x86": "Win32",
                 "x86_64": "x64",
             }[str(self.settings.arch)]
-            tools.mkdir(os.path.join(self.package_folder, "bin"))
+            mkdir(self, os.path.join(self.package_folder, "bin"))
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:104]:
```suggestion
```

File: recipes/yasm/all/conanfile.py [hunk 8]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
             build_type = "Debug" if self.settings.build_type == "Debug" else "Release"
-            shutil.copy(os.path.join(self._msvc_subfolder, arch, build_type, "yasm.exe"),
-                        os.path.join(self.package_folder, "bin", "yasm.exe"))
-            self.copy(pattern="yasm.exe*", src=self._source_subfolder, dst="bin", keep_path=False)
+            copy(self, pattern="yasm.exe",
+                    src=os.path.join(self._msvc_subfolder, arch, build_type),
+                    dst=os.path.join(self.package_folder, "bin"))
+            copy(self, pattern="yasm.exe*", src=self.source_folder, dst="bin", keep_path=False)
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:109]:
```suggestion
            copy(self, pattern="yasm.exe*", src=self.source_folder, dst=os.path.join(self.package_folder, "bin"), keep_path=False)
```

File: recipes/yasm/all/conanfile.py [hunk 9]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
         else:
-            autotools = self._configure_autotools()
+            autotools = Autotools(self)
             autotools.install()
-        tools.rmdir(os.path.join(self.package_folder, "share"))
+        rmdir(self, os.path.join(self.package_folder, "share"))
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:113]:
```suggestion
            rmdir(self, os.path.join(self.package_folder, "share"))
```

File: recipes/yasm/all/conanfile.py [hunk 10]
```diff
@@ -28,73 +28,91 @@ def configure(self):
...
 
     def package_info(self):
         bin_path = os.path.join(self.package_folder, "bin")
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:116]:
```suggestion
        self.cpp_info.includedirs = []
        self.cpp_info.libdirs = []
        
        bin_path = os.path.join(self.package_folder, "bin")
```

File: recipes/yasm/all/test_package/conanfile.py [hunk 11]
```diff
@@ -1,9 +1,25 @@
-from conans import ConanFile, tools
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.cmake import cmake_layout, CMake
+import os
```
Comment (SpaceIm) [inline recipes/yasm/all/test_package/conanfile.py:4]:
```suggestion
```

File: recipes/yasm/all/test_package/conanfile.py [hunk 12]
```diff
@@ -1,9 +1,25 @@
...
 
 
-class TestPackage(ConanFile):
+class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
+    generators = "CMakeDeps", "CMakeToolchain", "VirtualRunEnv"
```
Comment (SpaceIm) [inline recipes/yasm/all/test_package/conanfile.py:9]:
```suggestion
```

File: recipes/yasm/all/test_package/conanfile.py [hunk 13]
```diff
@@ -1,9 +1,25 @@
...
+    def requirements(self):
+        self.requires(self.tested_reference_str)
+
+    def layout(self):
+        cmake_layout(self)
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
```
Comment (SpaceIm) [inline recipes/yasm/all/test_package/conanfile.py:21]:
```suggestion
```

File: recipes/yasm/all/test_package/conanfile.py [hunk 14]
```diff
@@ -1,9 +1,25 @@
...
 
     def test(self):
-        if not tools.cross_building(self, skip_x64_x86=True):
+        if can_run(self):
             self.run("yasm --help", run_environment=True)
```
Comment (SpaceIm) [inline recipes/yasm/all/test_package/conanfile.py:25]:
```suggestion
            self.run("yasm --help", env="conanrun")
```

File: recipes/yasm/all/test_v1_package/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake
```
Comment (SpaceIm) [inline recipes/yasm/all/test_v1_package/conanfile.py:1]:
```suggestion
from conans import ConanFile
```

File: recipes/yasm/all/test_v1_package/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+from conan.tools.build import cross_building
+
+
+# legacy validation with Conan 1.x
```
Comment (SpaceIm) [inline recipes/yasm/all/test_v1_package/conanfile.py:5]:
```suggestion
```

File: recipes/yasm/all/test_v1_package/conanfile.py [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+class TestPackageV1Conan(ConanFile):
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "cmake", "cmake_find_package_multi"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
```
Comment (SpaceIm) [inline recipes/yasm/all/test_v1_package/conanfile.py:13]:
```suggestion
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/yasm/all/conanfile.py (hunk 5 above)
Comment (MartinDelille) [reply recipes/yasm/all/conanfile.py:12]: This is the one documented here: https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/autotools_package/all/conanfile.py#L11

How can I determine the minimum conan version this recipe is compatible with ?

File: recipes/yasm/all/conanfile.py (hunk 4 above)
Comment (MartinDelille) [reply recipes/yasm/all/conanfile.py]: Fixed here: 5851e48dd

**Comment by MartinDelille** [comment]
@madebr Any idea what is going wrong with the Windows build ?

**Comment by MartinDelille** [comment]
> I did something similar in the recipe for python 2.7.x

I can't find the recipe for python 2.7.x

**Comment by madebr** [comment]
Python is packaged as cpython.

**Comment by MartinDelille** [comment]
I can't find where the issue is. I'm not very familiar with sln project.

**Review by jwillikers — DISMISSED** [review]

File: recipes/yasm/all/conandata.yml [hunk 18]
```diff
@@ -4,3 +4,6 @@ sources:
       sha256: "3dce6601b495f5b3d45b59f7d2492a340ee7e84b5beca17e48f862502bd5603f"
     - url: "https://raw.githubusercontent.com/yasm/yasm/bcc01c59d8196f857989e6ae718458c296ca20e3/YASM-VERSION-GEN.bat"
       sha256: "b976cb97d2f7bb00e78e5db0da0978659acbdd60b52998cd2983145d8d75f141"
+patches:
+  "1.3.0":
+    - patch_file: "patches/0001-platformtoolset.patch"
```
Comment (prince-chrismc) [inline recipes/yasm/all/conandata.yml:9]: https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

File: recipes/yasm/all/conanfile.py [hunk 19]
```diff
@@ -1,9 +1,13 @@
-from conans import ConanFile, tools, AutoToolsBuildEnvironment, MSBuild
+from conan import ConanFile
+from conan.tools.files import apply_conandata_patches, chdir, copy, download, export_conandata_patches, get, rmdir
```
Comment (prince-chrismc) [inline recipes/yasm/all/conanfile.py:2]: linter

```suggestion
from conan.tools.files import apply_conandata_patches, copy, download, export_conandata_patches, get, rmdir
```

File: recipes/yasm/all/conanfile.py [hunk 20]
```diff
@@ -14,87 +18,90 @@ class YASMConan(ConanFile):
...
+            self._generate_autotools()
 
     def build(self):
-        if self.settings.compiler == "Visual Studio":
-            self._build_vs()
+        apply_conandata_patches(self)
+        if is_msvc(self):
+            msbuild = MSBuild(self)
+            msbuild.build_type = "Debug" if self.settings.build_type == "Debug" else "Release"
+            msbuild.platform = "Win32" if self.settings.arch == "x86" else msbuild.platform
```
Comment (prince-chrismc) [inline recipes/yasm/all/conanfile.py:76]: I am surprised these are not automatic 🤔 https://github.com/conan-io/conan/blob/dd2965c92da7bab30f3b483b54be24a9504ba4bb/conan/tools/microsoft/msbuild.py#L22\

If it's not working might be a bug

```suggestion
```

File: recipes/yasm/all/conanfile.py (hunk 5 above)
Comment (prince-chrismc) [reply recipes/yasm/all/conanfile.py:10]: theres a partial cheatsheet https://github.com/conan-io/conan-center-index/blob/master/docs/v2_linter.md#import-tools-from-conan

File: recipes/yasm/all/patches/0001-platformtoolset.patch [hunk 21]
```diff
@@ -0,0 +1,252 @@
+diff --git a/Mkfiles/vc10/genmacro/genmacro.vcxproj b/Mkfiles/vc10/genmacro/genmacro.vcxproj
+index 15cf1693..6c971f21 100644
+--- a/Mkfiles/vc10/genmacro/genmacro.vcxproj
++++ b/Mkfiles/vc10/genmacro/genmacro.vcxproj
+@@ -19,10 +19,12 @@
+   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'" Label="Configuration">
+     <ConfigurationType>Application</ConfigurationType>
+     <CharacterSet>MultiByte</CharacterSet>
++    <PlatformToolset>v142</PlatformToolset>
```
Comment (prince-chrismc) [inline recipes/yasm/all/patches/0001-platformtoolset.patch:9]: 🤔 https://github.com/conan-io/conan/blob/0ae90e8e63fef83486ef741a91108731450e028a/conan/tools/microsoft/toolchain.py#L38

The should be set by the new toolchain + build helper... maybe it's only the old compiler thats failing?

Comment (prince-chrismc) [reply recipes/yasm/all/patches/0001-platformtoolset.patch:9]: > A conanvcvars.bat file with the necessary vcvars invocation to define the build environment if necessary to build from the command line or from automated tools (might not be necessary if opening the IDE). This file will be automatically called by the tools.microsoft.MSBuild helper build() method.

https://docs.conan.io/en/latest/reference/conanfile/tools/microsoft.html#msbuildtoolchain

File: recipes/yasm/all/conandata.yml (hunk 18 above)
Comment (MartinDelille) [inline recipes/yasm/all/conandata.yml:9]:
```suggestion
    - patch_file: "patches/0001-platformtoolset.patch"
      patch_description: "Set plateformtoolset"
      patch_type: conan
```

Comment (MartinDelille) [reply recipes/yasm/all/conandata.yml:9]: Not sure if the fields are ok.

File: recipes/yasm/all/patches/0001-platformtoolset.patch (hunk 21 above)
Comment (MartinDelille) [reply recipes/yasm/all/patches/0001-platformtoolset.patch:9]: I don't know if this is the right way to do. I opened the `Mkfiles/vc10/yasm.sln` with Visual Studio 2019 and it asked me to upgrade the project which lead to this patch.

Maybe @madebr can help ?

I'm not familiar with MSBuild.

**Review by jwillikers — APPROVED** [review]

File: recipes/yasm/all/patches/0001-platformtoolset.patch (hunk 21 above)
Comment (prince-chrismc) [reply recipes/yasm/all/patches/0001-platformtoolset.patch:9]: https://github.com/conan-io/tribe/blob/main/design/032-msvc_support.md#motivation

Looking at the support is seems barely down to vc11 so I think the build helper just does not support this and slash or was never tested. In this release theres vc12 folder maybe we can try that?

If not in master they have newer vs build files we can use?

there's something missing with v2 if this needs to be added?

**Comment by uilianries** [comment]
Outdated Windows SDK? 🤔

**Comment by prince-chrismc** [comment]
No the upstream project structure is completely different for the newer version... I am playing with this local

should report back soon :)

**Comment by prince-chrismc** [comment]
I thought the new toolchain and build helper would support this **without** the patch but that is not the case. I made an issue to clarify how it works, https://github.com/conan-io/conan/issues/12314.

I played with it a bunch and it does the right thing (most) with the patch. The only draw back is if the toolset is none (which is default) it only compiles with 2019 toolset so that might burn someone 🙈 the old build helper did not really have that limitation.

***

My prescription for this PR

- revert to the vc10 build scripts
- apply the patch
- update the recipe with (so hopefully it's more obvious 🤞  whats going on when it fails)

```diff
diff --git a/recipes/yasm/all/conanfile.py b/recipes/yasm/all/conanfile.py
index 2c15c4c01..8f227335c 100644
--- a/recipes/yasm/all/conanfile.py
+++ b/recipes/yasm/all/conanfile.py
@@ -50,6 +50,8 @@ class YASMConan(ConanFile):

     def _generate_vs(self):
         tc = MSBuildToolchain(self)
+        if not self.settings.get_safe("compiler.toolset"):
+            tc.toolset = "v140"
         tc.generate()

     def _generate_autotools(self):
```

**Comment by madebr** [comment]
I got CMake to build yasm with MSVC: https://github.com/MartinDelille/conan-center-index/pull/10

**Review by jwillikers — APPROVED** [review]

File: recipes/yasm/all/conanfile.py [hunk 22]
```diff
@@ -14,87 +19,85 @@ class YASMConan(ConanFile):
...
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version][0],
-                  destination=self._source_subfolder, strip_root=True)
-        tools.download(**self.conan_data["sources"][self.version][1],
-                       filename=os.path.join(self._source_subfolder, "YASM-VERSION-GEN.bat"))
+        get(self, **self.conan_data["sources"][self.version][0],
+                  destination=self.source_folder, strip_root=True)
+        download(self, **self.conan_data["sources"][self.version][1],
+                       filename=os.path.join(self.source_folder, "YASM-VERSION-GEN.bat"))
```
Comment (madebr) [inline recipes/yasm/all/conanfile.py:46]: Is this still required?

**Review by jwillikers — APPROVED** [review]

File: recipes/yasm/all/conanfile.py [hunk 23]
```diff
@@ -14,87 +19,83 @@ class YASMConan(ConanFile):
...
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version][0],
-                  destination=self._source_subfolder, strip_root=True)
-        tools.download(**self.conan_data["sources"][self.version][1],
-                       filename=os.path.join(self._source_subfolder, "YASM-VERSION-GEN.bat"))
+        get(self, **self.conan_data["sources"][self.version][0],
+                  destination=self.source_folder, strip_root=True)
 
     @property
     def _msvc_subfolder(self):
```
Comment (madebr) [inline recipes/yasm/all/conanfile.py:47]: This property can be removed

Comment (MartinDelille) [reply recipes/yasm/all/conanfile.py]: done here: b4167c4c967058976885e8a0b4b4c2711c7823e1

File: recipes/yasm/all/conanfile.py (hunk 5 above)
Comment (MartinDelille) [reply recipes/yasm/all/conanfile.py:11]: The version requirement is good since we use `conan.tools.files.export_conandata_patches` introduced in 1.52.0.

**Review by jwillikers — APPROVED** [review]

File: recipes/yasm/all/test_package/conanfile.py [hunk 24]
```diff
@@ -1,9 +1,14 @@
-from conans import ConanFile, tools
+from conan import ConanFile
+from conan.tools.build import can_run
 
 
-class TestPackage(ConanFile):
+class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
+    test_type = "explicit"
```
Comment (uilianries) [inline recipes/yasm/all/test_package/conanfile.py:7]:
```suggestion
    generators = "VirtualRunEnv"
    test_type = "explicit"
```
It injects PATH

**Review by jwillikers — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
You'll need to retrigger CI, close the pr wait 10s and then re-open it 🔁

File: recipes/yasm/all/conanfile.py [hunk 25]
```diff
@@ -1,9 +1,14 @@
-from conans import ConanFile, tools, AutoToolsBuildEnvironment, MSBuild
+from conan import ConanFile
+from conan.tools.cmake import CMake, CMakeToolchain
+from conan.tools.files import apply_conandata_patches, copy, export_conandata_patches, get, rmdir
+from conan.tools.gnu import Autotools, AutotoolsToolchain
+from conan.tools.layout import basic_layout, cmake_layout
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:5]: `cmake_layout` should be imported from `conan.tools.cmake`

File: recipes/yasm/all/conanfile.py [hunk 26]
```diff
@@ -14,87 +19,79 @@ class YASMConan(ConanFile):
...
+    def export_sources(self):
+        export_conandata_patches(self)
+
     def configure(self):
         del self.settings.compiler.libcxx
         del self.settings.compiler.cppstd
 
-    def build_requirements(self):
-        if self._settings_build.os == "Windows" and self.settings.compiler != "Visual Studio" and not tools.get_env("CONAN_BASH_PATH"):
-            self.build_requires("msys2/cci.latest")
```
Comment (SpaceIm) [inline recipes/yasm/all/conanfile.py:33]: why was it removed?

**Pull request merged**